system-x/services/kudu/src/main/java/software/tnb/kudu/resource/openshift/OpenshiftKudu.java

```diff
@@ -0,0 +1,379 @@
+
+    final List<String> hostNames = List.of(MASTER_PREFIX, TSERVER_PREFIX);
+    private static final String KUDU_CLIENT = "kudu-client";
+
+    @Override
+    public void openResources() {
+        client = new RemoteClient(getMastersUsingRpc(), getClientPodName());
+        LOG.debug("master leader {}", client.findLeaderMasterServer());
+        createUIRoute();
+    }
+
+    private String getClientPodName() {
+        return OpenshiftClient.get().pods().withLabel(OpenshiftConfiguration.openshiftDeploymentLabel(), KUDU_CLIENT)
+            .list().getItems().get(0).getMetadata().getName();
+    }
+
+    @Override
+    public void closeResources() {
+        if (client != null) {
+            try {
+                client.close();
+                client = null;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Override
+    public void create() {
+        createServices();
+        createStatefulSet();
+        createClient();
+    }
+
+    @Override
+    public boolean isDeployed() {
+        return OpenshiftClient.get().getStatefulSet(MASTER_PREFIX) != null
+            && OpenshiftClient.get().getStatefulSet(TSERVER_PREFIX) != null
+            && servicePods().size() == getConfiguration().masterNumber() + getConfiguration().tabletNumber();
+    }
+
+    @Override
+    public Predicate<Pod> podSelector() {
+        return pod -> pod.getMetadata().getLabels().entrySet().stream()
+            .anyMatch(e -> e.getKey().equals(OpenshiftConfiguration.openshiftDeploymentLabel())
+            && (MASTER_PREFIX.equals(e.getValue()) || TSERVER_PREFIX.equals(e.getValue()) || KUDU_CLIENT.equals(e.getValue())));
+    }
+
+    @Override
+    public void undeploy() {
+        LOG.debug("Delete deployment {}", KUDU_CLIENT);
+        OpenshiftClient.get().apps().deployments().withName(KUDU_CLIENT).delete();
+
+        hostNames.forEach(statefulSet -> {
+            LOG.debug("Delete stateful set {}", statefulSet);
+            OpenshiftClient.get().apps().statefulSets().withName(statefulSet).delete();
+            WaitUtils.waitFor(() -> OpenshiftClient.get().getStatefulSet(statefulSet) == null, "Wait for stateful set "
+                + statefulSet + " is removed");
+        });
+
+        LOG.debug("Delete route with label {}={}", OpenshiftConfiguration.openshiftDeploymentLabel(), MASTER_PREFIX);
+        OpenshiftClient.get().routes().withLabel(OpenshiftConfiguration.openshiftDeploymentLabel(), MASTER_PREFIX).delete();
+
+        hostNames.forEach(s -> {
+            LOG.debug("Delete services with label {}={}", OpenshiftConfiguration.openshiftDeploymentLabel(), s);
+            OpenshiftClient.get().services().withLabel(OpenshiftConfiguration.openshiftDeploymentLabel(), s).delete();
+        });
+    }
+
+    @Override
+    public String externalHostname() {
+        return "http://" + OpenshiftClient.get().getRoute(createUIRoute()).getSpec().getHost();
```

Comment:
`externalHostname` should be where the client is connected. since you're using a custom pod as a client, you don't need this. you rather need `inClusterHostname` that returns something like `String.join(",", kuduCluster.getMastersUsingRpc())` as used in the test

Reply:
you are right, thanks